src/appleseed/renderer/modeling/postprocessingstage/vignettepostprocessingstage.cpp

```diff
@@ -94,45 +98,31 @@ namespace
             return true;
         }
 
-        void execute(Frame& frame) const override
+        void execute(Frame& frame, const std::size_t thread_count) const override
         {
+            // Skip vignetting if the intensity is zero.
+            if (feq(m_intensity, 0.0f))
+                return;
+
             const CanvasProperties& props = frame.image().properties();
-            const Vector2f resolution(static_cast<float>(props.m_canvas_width), static_cast<float>(props.m_canvas_height));
-            const Vector2f normalization_factor(lerp(resolution.y, resolution.x, m_anisotropy), resolution.y);
-
-            Image& image = frame.image();
-
-            for (std::size_t y = 0; y < props.m_canvas_height; ++y)
-            {
-                for (std::size_t x = 0; x < props.m_canvas_width; ++x)
-                {
-                    // Pixel coordinate normalized to be in the [-1, 1] range vertically.
-                    const Vector2f coord = (2.0f * Vector2f(static_cast<float>(x), static_cast<float>(y)) - resolution) / normalization_factor;
-
-                    //
-                    // Port of Keijiro Takahashi's natural vignetting effect for Unity.
-                    // Recreates natural illumination falloff, which is approximated by the "cosine fourth" law of illumination falloff.
-                    //
-                    // References:
-                    //
-                    //   https://github.com/keijiro/KinoVignette
-                    //   https://en.wikipedia.org/wiki/Vignetting#Natural_vignetting
-                    //
-
-                    const float linear_radial_falloff = norm(coord) * m_intensity;
-                    const float quadratic_radial_falloff = linear_radial_falloff * linear_radial_falloff + 1.0f;
-
-                    // Inversely proportional to the fourth power of the distance from the pixel to the image center.
-                    const float inverse_biquadratic_radial_falloff = 1.0f / (quadratic_radial_falloff * quadratic_radial_falloff);
-
-                    Color4f pixel;
-                    image.get_pixel(x, y, pixel);
-
-                    pixel.rgb() *= inverse_biquadratic_radial_falloff;
-
-                    image.set_pixel(x, y, pixel);
-                }
-            }
+
+            // Initialize effect-specific settings and context.
+            const VignetteParams effect_params {
+                m_intensity,
+                m_anisotropy,
+                static_cast<float>(props.m_canvas_width),
+                static_cast<float>(props.m_canvas_height)
+            };
+
+            // Instantiate an effect applier.
+            const VignetteApplierFactory effect_applier_factory(effect_params);
+            const std::unique_ptr<const IImageEffectApplier> effect_applier(effect_applier_factory.create());
+
+            // Apply the effect onto each image tile.
+            MultithreadPostProcessingStage::execute_on_tiles(
```

Comment:
I see now why you introduced `execute_on_tiles()` but IMHO you can simplify further: any post-processing stage based on `MultithreadPostProcessingStage` should mostly just pass the right effect (here the vignetting effect) to `MultithreadPostProcessingStage` and let that class handle everything.